File: 1101-1200/1135-Connecting Cities With Minimum Cost/1135-Connecting Cities With Minimum Cost.py
class Solution:
    def minimumCost(self, N: int, connections: List[List[int]]) -> int:
        parent = [i for i in range(1 + N)]
        def findParent(x):
            while parent[x] != x:
                parent[x] = parent[parent[x]]
                x = parent[x]
            return x
        
        connections.sort(key=lambda x: x[2])
        total = num = 0
        for city1, city2, cost in connections:
            p1 = findParent(city1)
            p2 = findParent(city2)
            if p1 != p2:
                parent[p2] = p1
                total += cost
                num += 1
                if num == N - 1:
                    return total
        
        return -1

File: 1101-1200/1135-Connecting Cities With Minimum Cost/1135-Connecting Cities With Minimum Cost.cpp
class Solution {
public:
    int minimumCost(int N, vector<vector<int>>& connections) {
        vector<int> parent(1 + N);
        for (int i = 1; i <= N; ++i) {
            parent[i] = i;
        }
        sort(connections.begin(), connections.end(), [](vector<int>& c1, vector<int>& c2){return c1[2] < c2[2];});
        int total = 0, num = 0;
        for (auto c : connections) {
            int p1 = findParent(parent, c[0]);
            int p2 = findParent(parent, c[1]);
            if (p1 != p2) {
                parent[p2] = parent[p1];
                total += c[2];
                if (++num == N - 1) {
                    return total;
                }
            }
        }
        
        return -1;  
    }
    
private:
    int findParent(vector<int>& parent, int x) {
        while (parent[x] != x) {
            parent[x] = parent[parent[x]];
            x = parent[x];
        }
        return x;
    }
};
